### PR TITLE
Add application relation data and event handling

### PR DESCRIPTION
**Pull request by dshcherb**
This change introduces app relation data handling based on the work done for Juju 2.7rc1.

### CODE REVIEW

File: juju/model.py [hunk 1]
```diff
@@ -168,26 +213,34 @@ def __getitem__(self, key):
...
+            return self._backend.relation_get(self.relation.id, self.entity.name, app=False)
         except RelationNotFound:
             # Dead relations tell no tales (and have no data).
             return {}
 
     def __setitem__(self, key, value):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
+            # This might happen due to a mid-hook leadership change besides trying to set data in remote data bags.
```
Comment (johnsca) [inline juju/model.py]: Can that really happen?  If so, ugh; I don't think any charm would handle that well and would just blow up.

File: juju/model.py [hunk 2]
```diff
@@ -43,32 +44,65 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
+
+    def _is_local_app(self, app_name):
+        return app_name == self.unit.app.name
```
Comment (niemeyer) [inline juju/model.py]: Per earlier PR, which had this embedded in it, these functions are not an improvement over the content they replace.

File: juju/model.py [hunk 3]
```diff
@@ -43,32 +44,65 @@ def get_relation(self, relation_name, relation_id=None):
...
+        return self._cache.get(CacheKey(Application, app_name), app_name, self._is_local_app(app_name), self._backend)
+
+
+class CacheKey:
+    def __init__(self, entity_type, *key_parts):
+        self.entity_type = entity_type
+        self.key_parts = key_parts
+
+    def __hash__(self):
+        return hash((self.entity_type, *self.key_parts))
```
Comment (niemeyer) [inline juju/model.py]: Why do we need the changes above?  Seems to be making things both more complex and less readable unnecessarily.

File: juju/model.py [hunk 4]
```diff
@@ -43,32 +44,65 @@ def get_relation(self, relation_name, relation_id=None):
...
+        self.app = cache.get(CacheKey(Application, app_name), app_name, is_local, backend)
+        self.is_local = is_local
+        self._backend = backend
+
+    @property
+    def is_leader(self):
+        if self.is_local:
+            # This value is not cached as it is not guaranteed to persist for the whole duration
+            # of a hook execution.
+            return self._backend.is_leader()
```
Comment (niemeyer) [inline juju/model.py]: How's that application data related?

**Comment by niemeyer** [comment]
Closing per the above comment. Let's please cut this down to smaller units so we can get to an agreement and merge them more easily.

**Pull request closed without merging**